core/src/main/clojure/xtdb/core/datalog.clj

```diff
@@ -510,32 +511,95 @@
-            [form-type
-             (case form-type
-               :logic-var (if (str/starts-with? (name form-arg) "?")
-                            form-arg
-                            (col-sym form-arg))
-               :fn-call (-> form-arg (update :args #(mapv with-col-metadata %)))
-               :literal form-arg)])]
+            (case form-type
+              :logic-var [:logic-var (if (str/starts-with? (name form-arg) "?")
+                                       form-arg
+                                       (col-sym form-arg))]
+              :fn-call [:fn-call (-> form-arg (update :args #(mapv with-col-metadata %)))]
+              :literal [:literal form-arg]
+              :sub-query [:logic-var (scalar-sub-query-placeholder form-arg)]))]
     (-> (s/unform ::form (with-col-metadata form))
-        (with-meta (into {::required-vars (form-vars form)}
+        (vary-meta merge
+                   {::required-vars (form-vars form)}
+                   (when-let [[return-type return-arg] return]
+                     (-> (case return-type
+                           :scalar {::return-var return-arg
+                                    ::vars #{return-arg}})
+                         (assoc ::return-type return-type)))))))
+
+(defn- ->apply-mapping [apply-params]
+  ;;TODO symbol names will clash with nested applies
+  ;; (where an apply is nested inside the dep side of another apply)
+  (when (seq apply-params)
+    (->> (for [param apply-params]
+           (let [param-symbol (-> (symbol (str "?ap_" param))
+                                  (with-meta {:correlated-column? true}))]
+             (MapEntry/create param param-symbol)))
+         (into {}))))
+
+(defn- plan-sub-query [{:keys [sub-query]}]
+  (let [required-vars (->> (:in sub-query)
+                           (into #{} (map
+                                       (fn [[in-type in-arg :as in]]
+                                         (when-not (= in-type :scalar)
+                                           (throw (err/illegal-arg :non-scalar-subquery-param
+                                                                   (s/unform ::in-binding in))))
+                                         in-arg))))
+        apply-mapping (->apply-mapping required-vars)]
+    (-> (plan-query (-> sub-query
+                        (dissoc :in)
+                        (assoc ::apply-mapping apply-mapping)))
+        (vary-meta into {::required-vars required-vars, ::apply-mapping apply-mapping}))))
 
-                         (when-let [[return-type return-arg] return]
-                           (-> (case return-type
-                                 :scalar {::return-var return-arg
-                                          ::vars #{return-arg}})
-                               (assoc ::return-type return-type))))))))
+(defn- wrap-scalar-sub-query [plan binding-sym scalar-sub-query param-vars]
+  (let [sq-plan (plan-sub-query scalar-sub-query)
```

Comment:
you might be able to save yourself the `:rename` if you `assoc :keys [binding-sym]` or similar?